src/sync/grandpa_warp_sync.rs

```diff
@@ -190,32 +194,56 @@ impl<TSrc> GrandpaWarpSync<TSrc> {
                         babe_next_epoch_query,
                         Some(current_epoch),
                         state,
+                        previous_error,
                     );
                 }
                 (
                     babe_fetch_epoch::Query::Finished {
                         result: Err(error),
-                        virtual_machine: _,
+                        virtual_machine,
                     },
-                    _,
-                ) => return Self::Finished(Err(Error::BabeFetchEpoch(error))),
-                (babe_fetch_epoch::Query::StorageGet(storage_get), fetched_current_epoch) => {
-                    return Self::InProgress(InProgressGrandpaWarpSync::StorageGet(StorageGet {
-                        inner: storage_get,
+                    fetched_current_epoch,
+                ) => {
+                    let babe_next_epoch_query =
+                        babe_fetch_epoch::babe_fetch_epoch(babe_fetch_epoch::Config {
+                            runtime: virtual_machine,
+                            epoch_to_fetch: if fetched_current_epoch.is_some() {
+                                babe_fetch_epoch::BabeEpochToFetch::NextEpoch
+                            } else {
+                                babe_fetch_epoch::BabeEpochToFetch::CurrentEpoch
+                            },
+                        });
+
+                    return return Self::from_babe_fetch_epoch_query(
```

Comment:
It's not mentioned in the guidelines but I'd strongly prefer to not use any sort of recursion.
Everything is already wrapped in a `loop`, so you should be able to update the local variables and continue looping.

Reply:
Also, it doesn't make sense to handle errors here this way.
All the parameters to the Babe epoch fetch are verified. The outcome is deterministic. If we try again, it's going to error again and again and again.

What we should do instead is start a warp sync request again (starting from the current state, not from where we were at the beginning), in the hopes that peers will give us another block.

For example, if we warp sync to block 1 million, and the runtime at block 1 million triggers a Babe fetch error, then we should try again making a request (starting from block 1 million) in the hopes that another peer will make us jump to block 2 million, where the runtime *is* compatible.